File: agents/memory/trace.py
"""Reasoning trace persistence — saves pipeline runs to MongoDB Atlas."""

import os
from datetime import datetime, timezone

from pymongo import MongoClient

_client: MongoClient | None = None


def _get_collection():
    global _client
    if _client is None:
        uri = os.environ.get("MONGODB_URI")
        if not uri:
            raise EnvironmentError("MONGODB_URI is not set")
        _client = MongoClient(uri)
    db_name = os.environ.get("MONGODB_DB_NAME", "atlas")
    return _client[db_name]["reasoning_traces"]


def save_trace(
    ticker: str,
    user_id: str,
    boundary_mode: str,
    technical: dict,
    fundamental: dict,
    sentiment: dict,
    synthesis: dict,
    risk: dict,
    final_decision: dict,
) -> str:
    doc = {
        "ticker": ticker,
        "user_id": user_id,
        "boundary_mode": boundary_mode,
        "created_at": datetime.now(timezone.utc),
        "pipeline_run": {
            "technical": technical,
            "fundamental": fundamental,
            "sentiment": sentiment,
            "synthesis": synthesis,
            "risk": risk,
            "final_decision": final_decision,
        },
    }
    result = _get_collection().insert_one(doc)
    return str(result.inserted_id)

File: frontend/lib/agents/memory/trace.ts
/**
 * Reasoning trace persistence — saves pipeline runs to MongoDB Atlas.
 *
 * Mirrors backend/agents/memory/trace.py exactly.
 *
 * Collection: atlas.reasoning_traces
 * Schema:
 *   ticker, user_id, boundary_mode, created_at,
 *   pipeline_run: { technical, fundamental, sentiment, synthesis, risk, final_decision }
 */

import { MongoClient } from "mongodb";
import type { TechnicalOutput, FundamentalOutput, SentimentOutput, SynthesisOutput, RiskOutput, PortfolioDecision } from "../state";

let _client: MongoClient | null = null;

function getCollection() {
  if (!_client) {
    const uri = process.env.MONGODB_URI;
    if (!uri) {
      throw new Error("MONGODB_URI is not set");
    }
    _client = new MongoClient(uri);
  }
  const dbName = process.env.MONGODB_DB_NAME ?? "atlas";
  return _client.db(dbName).collection("reasoning_traces");
}

export interface SaveTraceParams {
  ticker: string;
  userId: string;
  boundaryMode: string;
  technical: TechnicalOutput | Record<string, unknown>;
  fundamental: FundamentalOutput | Record<string, unknown>;
  sentiment: SentimentOutput | Record<string, unknown>;
  synthesis: SynthesisOutput | Record<string, unknown>;
  risk: RiskOutput | Record<string, unknown>;
  finalDecision: PortfolioDecision | Record<string, unknown>;
}

/**
 * Persist a full pipeline run to MongoDB.
 *
 * @returns Inserted document _id as a string.
 */
export async function saveTrace(params: SaveTraceParams): Promise<string> {
  const {
    ticker,
    userId,
    boundaryMode,
    technical,
    fundamental,
    sentiment,
    synthesis,
    risk,
    finalDecision,
  } = params;

  const doc = {
    ticker,
    user_id: userId,
    boundary_mode: boundaryMode,
    created_at: new Date(),
    pipeline_run: {
      technical,
      fundamental,
      sentiment,
      synthesis,
      risk,
      final_decision: finalDecision,
    },
  };

  const collection = getCollection();
  const result = await collection.insertOne(doc);
  return result.insertedId.toString();
}
